feat(platform): add documentation for scoped API tokens

tobice opened the pull request:
We're adding public facing documentation for a new feature, tokens with limited permissions, aka scoped tokens.

Full context is available in the [design doc](https://www.notion.so/apify/API-permissions-for-public-Actors-d58ddbd3ec8849a1ad9940dc8a556e87#c91110201b2b4ee7aaeab57c761eb16a).

tobice commented:
Hey folks, asking for early review. What's missing are screenshots, which I won't be able to provide before we finish the UI. However, the copy is final (= ready for review) from my PoV.

Also tagged @jirimoravcik as he is only semi-familiar with the project, so he'll be able to give useful feedback if the documentation is easy to understand / sufficiently detailed.

fnesveda commented on sources/platform/integrations/programming/api.md:
```diff
@@ -45,3 +51,58 @@ When working under an organization account, you will see two types of API tokens
+This way you can be sure that once you give a token the permission to run an Actor, it will just work, and you don't have to worry
+about the exact permissions the Actor might need. However, this also means that you need to trust the Actor.
+
+:::info Note
```
It is so confusing that `:::info` has a !⃝ icon, but `:::note` has a ⓘ icon 🤯

tobice replied:
I actually did not notice this :smile: Yeah, it's weird... unified it to use `note` across the board. It's less scary.

fnesveda reviewed (changes requested):
Let's put in some temporary screenshots (either the current state, or screenshots from Figma) until we finish the full UI, so that we can already craft the docs around them. Currently the docs are missing instructions on how to create scoped tokens, and what all the various permissions are.

I also think the API page should be displayed more prominently, not just two levels deep in Integrations, where it gets a bit lost. But that's a problem for another task.

jirimoravcik reviewed (approved):
Looks good to me, I'll also check when the screenshots are added

tobice commented:
**Updated the documentation:**

- Added short instructions on how to create a scoped token
- Added 2 temporary screenshots, to be replaced with the actual ones once the UI is finished (I see no other use cases for screenshots)
- Unified callouts to use `note` across the board
- Minor copy changes

tobice commented:
> Currently the docs are missing [...], and what all the various permissions are.

Agreed with Franta that this is not necessary.

> I also think the API page should be displayed more prominently, not just two levels deep in Integrations, where it gets a bit lost. But that's a problem for another task.

Hmm I guess. cc @TC-MO

fnesveda reviewed (approved):
👏 Good job, great docs 👍

TC-MO commented on sources/platform/integrations/programming/api.md:
```diff
@@ -24,12 +24,18 @@ To access the Apify API in your integrations, you need to authenticate using you
 
-> **IMPORTANT**: **Do not share the API token with untrusted parties, or use it directly from client-side code,
-unless you fully understand the consequences!**
+:::caution Important
+Do not share the API token with untrusted parties, or use it directly from client-side code,
+unless you fully understand the consequences!
+:::
```
```suggestion
:::caution Secure token usage

Never share your API token with untrusted parties or use it directly in client-side code unless you fully understand the security risks.

:::
```

TC-MO commented on sources/platform/integrations/programming/api.md:
```diff
@@ -40,8 +46,73 @@ do not require an authentication token because they contain a hard-to-guess iden
+
+![An example scoped token that combines account level permissions and resource-specific permissions](../images/tmp__scoped-token-combining-permissions.png)
+
+:::note Note
+Keep in mind that the scoped tokens still operate within the context of your account. A scoped token will never be able to access resources that your account doesn't have access to.
+:::
```
```suggestion
:::note Token scope limits

Scoped tokens operate within the context of your account and cannot access resources beyond your account's permissions.

:::
```

tobice replied:
What's the reasoning here? "Keep in mind" was meant to acknowledge the fact that this is actually to some extent obvious ("goes without saying"), so it is basically just a reminder to the reader how this works.

TC-MO replied:
Keep in mind from my POV seems like unnecessary fluff. In admonitions I try to go straight to the point and not make any assumptions about knowledge of reader. To some it might be obvious but not all.

tobice replied:
Okay, in that case I'd probably remove this callout completely. It doesn't bring much value anyway.

tobice replied:
Removed.

TC-MO commented on sources/platform/integrations/programming/api.md:
```diff
@@ -40,8 +46,73 @@ do not require an authentication token because they contain a hard-to-guess iden
+
+Once you create a new resource with the token, **the token will gain full access to that resource**, regardless of other permissions. It is not possible to create a token that can create a dataset, but not write to it.
+
+:::tip
+This is useful if you want to for example create a token that can dynamically create & populate datasets, but without the need to access other datasets in your account.
+:::
```
```suggestion
:::tip Selective token permissions

This is useful if you want to for example create a token that can dynamically create & populate datasets, but without the need to access other datasets in your account.

:::
```

TC-MO commented on sources/platform/integrations/programming/api.md:
```diff
@@ -40,8 +46,73 @@ do not require an authentication token because they contain a hard-to-guess iden
+This way you can be sure that once you give a token the permission to run an Actor, it will just work, and you don't have to worry
+about the exact permissions the Actor might need. However, this also means that you need to trust the Actor.
+
+:::note Note
+This is why **we currently do not allow scoped tokens to create or modify Actors**. With those permissions it would be easy for the token to upload malicious code and gain access to your full account. If you do need to create or modify Actors via Apify API, you need to use an unscoped token.
+:::
```
```suggestion
:::note Actor modification restriction 

For security reasons, scoped tokens cannot create or modify Actors. Use an unscoped token for these operations.

:::
```

tobice replied:
The idea here is to make it less wordy? :thinking: 

I'd still like to link the statement here to the previous paragraph.

TC-MO commented on sources/platform/integrations/programming/api.md:
```diff
@@ -40,8 +46,73 @@ do not require an authentication token because they contain a hard-to-guess iden
+
+Therefore, you need to make sure the token has sufficient permissions not only to set up the webhook, but also to perform the actual operation.
+
+:::tip Example
+Let's say you want to create a webhook that pushes an item to a dataset every time an Actor successfully finishes. Then such a scoped token needs to be allowed to both run the Actor (in order to create the webhook), and write to that dataset.
+:::
```
```suggestion
:::note Multi-permission example

Let's say you want to create a webhook that pushes an item to a dataset every time an Actor successfully finishes. Then such a scoped token needs to be allowed to both run the Actor (in order to create the webhook), and write to that dataset.

:::
```

TC-MO commented on sources/platform/integrations/programming/api.md:
```diff
@@ -40,8 +46,73 @@ do not require an authentication token because they contain a hard-to-guess iden
+
+- **Resource-specific permissions**: These will apply only to specific, existing resources. For example, you can use these to allow the token to read from a particular dataset.
+
+:::tip
+A single token can combine both types. You can create a token that can _read_ any data storage, but _write_ only to one specific dataset.
+:::
```
```suggestion
:::tip Combined token permissions

A single token can combine both types. You can create a token that can _read_ any data storage, but _write_ only to one specific dataset.

:::
```

tobice replied:
Are these callout titles a strong requirement?

TBH, I don't see that much value in them... ideally there'd be no title, just the text plus the icon, but since the title has to be there, for me "Tip" / "Note" / "Important" work better. The proposed titles seem very vague and non-informative to me.

TC-MO replied:
In my experience documentation is more scanned than read. These admonitions and titles within them, can help grab attention as well as at a glance help user decide whether he found what was looking for or not. Moreover that is also how we are doing them anywhere else in docs. I would be in favour of keeping the title, but can definitely come up with few alternatives if these seem to vague

tobice replied:
> In my experience documentation is more scanned than read.

Totally on board with this :+1: 

But IMHO the generic titles such as "important" or "tip" can actually greatly improve scannability: If something is a tip, I know I can skip it, if it's important, I should probably read it etc.

My argument is basically that these generic titles might bring more value than callout-specific, yet non-informative titles.

Btw Docusaurus itself [combines both approaches](https://docusaurus.io/docs/next/markdown-features/toc). Specific titles where it makes sense, generic where it doesn't.

I'll also try to think about alternatives for specific titles, but I'd actually prefer to keep the generic ones as fallback.

TC-MO commented on sources/platform/integrations/programming/api.md:
```diff
@@ -24,12 +24,18 @@ To access the Apify API in your integrations, you need to authenticate using you
 
 ![Integrations page in Apify Console](../images/api-token.png)
 
+:::caution Warning
+By default, tokens can access all data in your account. If you plan to share the token with a 3rd party, consider [limiting the permission scope](#limited-permissions) of the token, so that it can only access what it really needs.
+:::
```
```suggestion
:::caution Token access scope

By default, tokens have full access to your account data. For third-party integratns, consider [limiting the token's permissions](#limited-permissions).

:::
```

tobice commented:
Cool, thanks for the review :heart: 

Let's wait now for Kuba and for how we decide to proceed with the UI.

TC-MO commented on sources/platform/integrations/programming/api.md:
```diff
@@ -40,8 +46,73 @@ do not require an authentication token because they contain a hard-to-guess iden
+This is useful if you want to for example create a token that can dynamically create & populate datasets, but without the need to access other datasets in your account.
+:::
+
+### Running and scheduling Actors
```
```suggestion
### Run and schedule Actors
```

tobice replied:
Same as above, this is more _how running and scheduling Actors work_ than _how to run and schedule Actors_.

TC-MO commented on sources/platform/integrations/programming/api.md:
```diff
@@ -40,8 +46,73 @@ do not require an authentication token because they contain a hard-to-guess iden
+This is why **we currently do not allow scoped tokens to create or modify Actors**. With those permissions it would be easy for the token to upload malicious code and gain access to your full account. If you do need to create or modify Actors via Apify API, you need to use an unscoped token.
+:::
+
+### Setting up webhooks
```
```suggestion
### Set up webhooks
```

tobice replied:
Dtto as above.

TC-MO commented on sources/platform/integrations/programming/api.md:
```diff
@@ -40,8 +46,73 @@ do not require an authentication token because they contain a hard-to-guess iden
+Keep in mind that the scoped tokens still operate within the context of your account. A scoped token will never be able to access resources that your account doesn't have access to.
+:::
+
+### Allowing tokens to create resources
```
```suggestion
### Allow tokens to create resources
```

tobice replied:
Hmmm here the idea is not to _allow the token to create resource_, but rather _how it works when you allow tokens to create resources_.

If the gerund is a problem, I'd go for something like "How creating resources with scoped tokens work"... but that's a tad long and awkward.

TC-MO commented on sources/platform/integrations/programming/api.md:
```diff
@@ -40,8 +46,73 @@ do not require an authentication token because they contain a hard-to-guess iden
+
+If you need to create new resources with the token (for example, create a new Task, or storage), you need to explicitly allow that as well.
+
+Once you create a new resource with the token, **the token will gain full access to that resource**, regardless of other permissions. It is not possible to create a token that can create a dataset, but not write to it.
```
```suggestion
Once you create a new resource with the token, _the token will gain full access to that resource_, regardless of other permissions. It is not possible to create a token that can create a dataset, but not write to it.
```

TC-MO reviewed (changes requested):
Added a few changes to notes & headings

drobnikj commented:
Changes look good, let's wait to final design implementation.

tobice commented:
Alright, this PR is pretty much ready.

We synced with @TC-MO and made some changes together:

- We got rid of some of the gerunds in headings
- We kept the generic admotion titles ("Tip" / "Note") to see if people like them

I resolved all the pending conversions which were related to these topics.

@TC-MO @drobnikj feel free to take a look and approve :pray:

drobnikj reviewed (approved):
🆒

drobnikj commented:
I merged [the public release](https://github.com/apify/apify-core/pull/17354) to develop today, we can manage it to prod today.
Let's merge this @tobice

tobice commented:
Won't make it to today's release. Lets's merge it tomorrow.
